src/sagemaker/estimator.py

```diff
@@ -875,27 +901,34 @@ def _stage_user_code_in_s3(self):
 
         if self.code_location is None and local_mode:
             code_bucket = self.sagemaker_session.default_bucket()
-            code_s3_prefix = '{}/source'.format(self._current_job_name)
+            code_s3_prefix = '{}/{}'.format(self._current_job_name, s3_location)
             kms_key = None
 
         elif self.code_location is None:
             code_bucket, _ = parse_s3_url(self.output_path)
-            code_s3_prefix = '{}/source'.format(self._current_job_name)
+            code_s3_prefix = '{}/{}'.format(self._current_job_name, s3_location)
             kms_key = self.output_kms_key
         else:
             code_bucket, key_prefix = parse_s3_url(self.code_location)
-            code_s3_prefix = '/'.join(filter(None, [key_prefix, self._current_job_name, 'source']))
+            code_s3_prefix = '/'.join(filter(None, [key_prefix, self._current_job_name, s3_location]))
 
             output_bucket, _ = parse_s3_url(self.output_path)
             kms_key = self.output_kms_key if code_bucket == output_bucket else None
 
-        return tar_and_upload_dir(session=self.sagemaker_session.boto_session,
-                                  bucket=code_bucket,
-                                  s3_key_prefix=code_s3_prefix,
-                                  script=self.entry_point,
-                                  directory=self.source_dir,
-                                  dependencies=self.dependencies,
-                                  kms_key=kms_key)
+        if tar:
+            return tar_and_upload_dir(session=self.sagemaker_session.boto_session,
+                                      bucket=code_bucket,
+                                      s3_key_prefix=code_s3_prefix,
+                                      script=self.entry_point,
+                                      directory=self.source_dir,
+                                      dependencies=self.dependencies,
+                                      kms_key=kms_key)
+        else:
+            return upload_file(session=self.sagemaker_session.boto_session,
```

Comment:
We agreed here to only upload compressed files.

Reply:
+1 will send out a change to reflect that